homeassistant/components/light/group.py

```diff
@@ -0,0 +1,280 @@
+
+    async def async_added_to_hass(self) -> None:
+        """Register callbacks."""
+        @callback
+        def async_state_changed_listener(entity_id: str, old_state: State,
+                                         new_state: State):
+            """Handle child updates."""
+            self.async_schedule_update_ha_state(True)
+
+        self._async_unsub_state_changed = async_track_state_change(
+            self.hass, self._entity_ids, async_state_changed_listener)
+
+    async def async_will_remove_from_hass(self):
+        """Callback when removed from HASS."""
+        if self._async_unsub_state_changed:
+            self._async_unsub_state_changed()
+            self._async_unsub_state_changed = None
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def state(self) -> str:
+        """Return the state."""
+        return self._state
+
+    @property
+    def is_on(self) -> bool:
+        """Return True if entity is on."""
+        return self._state == STATE_ON
+
+    @property
+    def brightness(self) -> Optional[int]:
+        """Return the brightness of this light between 0..255."""
+        return self._brightness
+
+    @property
+    def xy_color(self) -> Optional[Tuple[float, float]]:
+        """Return the XY color value [float, float]."""
+        return self._xy_color
+
+    @property
+    def rgb_color(self) -> Optional[Tuple[int, int, int]]:
+        """Return the RGB color value [int, int, int]."""
+        return self._rgb_color
+
+    @property
+    def color_temp(self) -> Optional[int]:
+        """Return the CT color value in mireds."""
+        return self._color_temp
+
+    @property
+    def min_mireds(self) -> Optional[int]:
+        """Return the coldest color_temp that this light supports."""
+        return self._min_mireds
+
+    @property
+    def max_mireds(self) -> Optional[int]:
+        """Return the warmest color_temp that this light supports."""
+        return self._max_mireds
+
+    @property
+    def white_value(self) -> Optional[int]:
+        """Return the white value of this light between 0..255."""
+        return self._white_value
+
+    @property
+    def effect_list(self) -> Optional[List[str]]:
+        """Return the list of supported effects."""
+        return self._effect_list
+
+    @property
+    def effect(self) -> Optional[str]:
+        """Return the current effect."""
+        return self._effect
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return self._supported_features
+
+    @property
+    def should_poll(self) -> bool:
+        """No polling needed for a group light."""
+        return False
+
+    async def _async_send_message(self, service, **kwargs):
+        """Send a message to all entities in the group."""
+        for entity_id in self._entity_ids:
+            payload = deepcopy(kwargs)
```

Comment:
No need to do a deepcopy if you only mutate the outer layer. Just `dict(kwargs)` is enough.

Also, are all kwargs for `turn_on` and `turn_off` the same as service attributes? This becomes an implicit rule with this method and I'm not sure if this is good.

Reply:
I mean as far as I saw, they're the same, but you're right in that we shouldn't rely on that. So I now changed the code to use [`light.async_turn_on`](https://github.com/home-assistant/home-assistant/blob/9658f4383cd391145b4fb23f135437345eb3abba/homeassistant/components/light/__init__.py#L163-L188) directly. I mean then we also need to rely on *that method* having the same kwargs, but that's much more maintainable.

We could also manually specify the arguments in our own `self.async_turn_on`, but that'd be a *lot* of arguments (and a bit ugly 😅).

Reply:
Problem with `light.async_turn_on` is that it's non blocking (see other comment)